[451] Refactoring metrics docs (#452)
tarantool/metrics: doc/monitoring/getting_started.rst

..  _monitoring-getting_started:

Getting started with monitoring
===============================

If you use Tarantool version below `2.11.1 <https://github.com/tarantool/tarantool/releases/tag/2.11.1>`__,
it is necessary to install the latest version of ``metrics`` first. For details,
see :ref:`Installing the metrics module <install>`.

..  _monitoring-getting_started-usage:

Using the metrics module
------------------------

..  note::

    The module is also used in applications based on the Cartridge framework. For details,
    see the :ref:`Getting started with Cartridge application <metrics-getting_started_cartridge>` section.

#.  First, set the instance name and start to collect the standard set of metrics.

>     Also, you can set a global label for your instance.

Reviewer: "set the instance name" and "you can set a global label for your instance" is the same action, so info is duplicated here.

Reviewer 2: Fixed

    ..  code-block:: lua

>         metrics.cfg(labels = {alias = 'my-instance'})

Reviewer: ```suggestion
        metrics.cfg{labels = {alias = 'my-instance'}}
```
(Otherwise, it's an invalid expression in terms of syntax.)

Reviewer 2: Fixed

        },
        "timestamp": 1679663602823779,
        "metric_name": "tnt_vinyl_disk_index_size",
        "value": 0
      },
      {
        "label_pairs": {
          "alias": "my-instance"
        },
        "timestamp": 1679663602823779,
        "metric_name": "tnt_info_memory_data",
        "value": 39272
      },
      {
        "label_pairs": {
          "alias": "my-instance"
        },
        "timestamp": 1679663602823779,
        "metric_name": "tnt_election_vote",
        "value": 0
      }
    ]

The data can be visualized in
`Grafana dashboard <https://www.tarantool.io/en/doc/latest/book/monitoring/grafana_dashboard/#monitoring-grafana-dashboard-page>`__.

The full source example is listed below:

..  code-block:: lua

    -- Import modules
    local metrics = require('metrics')
    local http_server = require('http.server')
    local json_exporter = require('metrics.plugins.json')

    -- Define helper functions
    local function http_metrics_handler(request)
        return request:render({ text = json_exporter.export() })
    end

    -- Start the database
    box.cfg{
        listen = 3301,
    }

    -- Configure the metrics module

>     metrics.cfg(labels = {alias = 'my-tnt-app'})

Reviewer: ```suggestion
    metrics.cfg{labels = {alias = 'my-tnt-app'}}
```

    -- Run the web server
    local server = http_server.new('0.0.0.0', 8081)
    server:route({path = '/metrics'}, http_metrics_handler)
    server:start()

..  _monitoring-getting_started-http_metrics:

Collecting HTTP metrics
-----------------------

To enable the collection of HTTP metrics, wrap a handler with a ``metrics.http_middleware.v1`` function:

..  code-block:: lua

    local metrics = require('metrics')
    local httpd = require('http.server').new(ip, port)

    -- Create a summary collector for latency

>     local default_collector = metrics.http_middleware.build_default_collector('summary')

Reviewer: You can use single `configure_default_collector` instead of `build`+`set`

Reviewer 2: Fixed

    metrics.http_middleware.set_default_collector(default_collector)

    -- Set a route handler for latency summary collection

>     httpd:route({ path = '/path-1', method = 'POST' }, metrics.http_middleware.v1(handler_1, collector))

Reviewer: ```suggestion
    httpd:route({ path = '/path-1', method = 'POST' }, metrics.http_middleware.v1(handler_1))
```
After snippet update, there are no `collector` variable. You should either pass `, default_collector` or do not pass anything and use defaults (which is `default_collector` now).

Reviewer 2: Fixed

Reviewer: It seems that changes haven't been pushed yet.

Reviewer 2: Pushed the following changes + removed tarantoolctl from the instructions

Monitoring: getting started
===========================

.. _monitoring-getting_started-install:

Install metrics rock
-----------------------

Install the latest version of Metrics. For instructions to install the Metrics

> rock :ref:`view the install documentation <_install-the_usual_way>`.

Author: Дописал и тут и там

.. _monitoring-getting_started-how_to_use:

How to use
---------------

.. hint::

>     Использование модуля совместно с фреймворком Cartridge рассмотрено

Author: Поправил

    :ref:`здесь <_getting_started_cartridge>`.

1. Устанавливаем название узла и включаем вывод стандартного набора метрик
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You may also set a global label for your instance:

Author: Добавил

Monitoring: getting started
===========================

.. _monitoring-getting_started-install:

Install metrics rock
-----------------------

> If you are using Tarantool version below 2.11.1 - install the latest version of Metrics. For instructions

Reviewer: Мне кажется, мы нигде не используем название модуля как "Metrics" с большой буквы. Давай везде с маленькой писать, в кавычках для форматирования, как было до изменений

to install the Metrics rock :ref:`view the install documentation <_install-the_usual_way>`.

.. _monitoring-getting_started-how_to_use:

How to use
---------------

.. hint::

    The use of the module in applications based on the Cartridge framework

>     is considered :ref:`here <_getting_started_cartridge>`.

Reviewer: ```suggestion
    is described :ref:`here <_getting_started_cartridge>`.
```

> 1. Set the node name and enable the output of the standard set of metrics

Reviewer: ```suggestion
1. Set the instance name and start to collect the standard set of metrics
```

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
```
Don't use the heading formatting with the lists. If you need to wrap some text as the list item, use the list indentation instead: https://www.tarantool.io/en/doc/latest/contributing/docs/markup/intro/#indentation

Calling ``metrics.cfg()`` enables collection of a standard set of metrics.
You may also set a global label for your instance:

..  code-block:: lua

    metrics.cfg(labels = {alias = 'my-instance'})

> .. hint::

Reviewer 2: ```suggestion
..  note::
```

>     When using a metrics module version below 0.17.0, instead of using the function ``metrics.cfg(...)``

Reviewer: ...use the following snippet instead of ``metrics.cfg(...)``

    you should write it like this:

    ..  code-block:: lua

        metrics.set_global_labels({alias = 'my-instance'})
        metrics.enable_default_metrics()

2. Adding a handler for the command to retrieve metric values
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> * If you need the JSON format:

Reviewer: ```suggestion
* For JSON format:
```

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> * If you need the JSON format:

Reviewer 2: ```suggestion

    For JSON format:
```
Also, don't forget to add list indentation for the code-blocks below

..  code-block:: lua

    local json_exporter = require('metrics.plugins.json')
    local function http_metrics_handler(request)
        return request:render({ text = json_exporter.export() })
    end

> * If you need the Prometheus format:

Reviewer: То же самое

..  code-block:: lua

    local prometheus_exporter = require('metrics.plugins.prometheus').collect_http

> * To learn how to obtain custom metrics, check the :ref:`API reference <metrics-api_reference>`.

Reviewer: ```suggestion
* To learn how to extend metrics with custom data, check the :ref:`API reference <metrics-api_reference>`.
```

> 3. Registering the command to retrieve metric values and starting the HTTP server

Reviewer: ```suggestion
3. Starting the HTTP server and expose metrics
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

..  code-block:: lua

    local http_server = require('http.server')
    local server = http_server.new('0.0.0.0', 8081)
    server:route({path = '/metrics'}, http_metrics_handler)
    server:start()

.. _monitoring-getting_started-result_example:

> Result example

Reviewer: ```suggestion
Full example
```

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> ..  code-block:: lua
> 
>     local http_server = require('http.server')
>     local server = http_server.new('0.0.0.0', 8081)
>     server:route({path = '/metrics'}, http_metrics_handler)
>     server:start()
> 
> .. _monitoring-getting_started-result_example:
> 
> Result example

Reviewer 2: ```suggestion

    ..  code-block:: lua

        local http_server = require('http.server')
        local server = http_server.new('0.0.0.0', 8081)
        server:route({path = '/metrics'}, http_metrics_handler)
        server:start()

    .. _monitoring-getting_started-result_example:

    Full example
```

---------------

> In the end, you will be able to see the metric values by accessing the URL http://localhost:8081/metrics:

Reviewer 2: ```suggestion
In the end, you will be able to see the metric values by accessing the URL ``http://localhost:8081/metrics``:
```

      {
        "label_pairs": {
          "alias": "my-instance"
        },
        "timestamp": 1679663602823779,
        "metric_name": "tnt_election_vote",
        "value": 0
      }
    ]

The data can be visualized in
`Grafana dashboard <https://www.tarantool.io/en/doc/latest/book/monitoring/grafana_dashboard/#monitoring-grafana-dashboard-page>`__.

.. _monitoring-getting_started-full_source_example:

Full source example
-------------------

.. code-block:: lua

    -- Import modules
    local metrics = require('metrics')
    local http_server = require('http.server')
    local json_exporter = require('metrics.plugins.json')

    -- Define helper functions
    local function http_metrics_handler(request)
        return request:render({ text = json_exporter.export() })
    end

    -- Start the database
    box.cfg{
        listen = 3301,
    }

    -- Configure the metrics module
    metrics.cfg(labels = {alias = 'my-tnt-app'})

    -- Run the web server
    local server = http_server.new('0.0.0.0', 8081)
    server:route({path = '/metrics'}, http_metrics_handler)
    server:start()

..  _monitoring-getting_started-http_metrics:

Collect HTTP metrics
--------------------

> To enable the collection of HTTP metrics, you need wrap handler to function

Reviewer: ```suggestion
To enable the collection of HTTP metrics, you need wrap handler with a function
```

Monitoring: getting started
===========================

.. _monitoring-getting_started-install:

Install metrics rock
-----------------------

Install the latest version of Metrics. For instructions to install the Metrics
rock :ref:`view the install documentation <_install-the_usual_way>`.

.. _monitoring-getting_started-how_to_use:

How to use
---------------

.. hint::

    Использование модуля совместно с фреймворком Cartridge рассмотрено
    :ref:`здесь <_getting_started_cartridge>`.

1. Подключаем модули в коде приложения
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

..  code-block:: lua

    local http_server = require('http.server')
    local metrics = require('metrics')

> 2. Добавляем обработчик команды

Reviewer: Какой команды? Зачем это делать в первую очередь после подключения модуля? Задание экспортера лучше переместить туда, где его собираются использовать -- в пункт с `server:route`. `require('http.server')` тоже уместно переместить туда.

Author: Переписал. Довольно большой блок получится, если их смешать. Придвинул ближе туда. Дописал название:
Добавляем обработчик команды получения значений метрик

Reviewer: Например, в случае Prometheus "блок" -- это лишняя строчка по перекладыванию функции из одного места в другое. Даже если очень хочется разделить их на два блока, имеет смысл положить эти два блока друг рядом с другом. А вот подключать `require('http.server')` при `require('metrics')` не несёт никакой осмысленной нагрузки.

Author: Убрал первый пункт

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Если нужен формат JSON:

..  code-block:: lua

    local json_exporter = require('metrics.plugins.json')
    local function http_metrics_handler(request)
        return request:render({ text = json_exporter.export() })
    end

* Если нужен формат Prometheus:

..  code-block:: lua

    local prometheus_exporter = require('metrics.plugins.prometheus')

>     local function http_metrics_handler(...)

Reviewer: Эта функция не имеет никакого смысла. Если очень хочется переложить, можно просто сделать `local http_metrics_handler = require('metrics.plugins.prometheus')`

Author: Оно же так не работает:
``` lua
local metrics = require('metrics')
local http_server = require('http.server')

local http_metrics_handler = require('metrics.plugins.prometheus') -- <<<<< --

local server = http_server.new('0.0.0.0', 8081)
server:route({path = '/metrics'}, http_metrics_handler) -- << -- wrong argument: expected function, but received table
server:start()
```

```
LuajitError: .../projects/metrics/.rocks/share/tarantool/http/server.lua:7: wrong argument: expected function, but received table
```

Author: Может ты имел ввиду вот так?
``` lua
local http_metrics_handler = require('metrics.plugins.prometheus').collect_http
```

Author: Поправил

        return prometheus_exporter.collect_http(...)
    end

* To learn how to obtain custom metrics, check the :ref:`API reference <metrics-api_reference>`.

3. Устанавливаем название узла
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You may also set a global label for your instance:

..  code-block:: lua

    metrics.set_global_labels{alias = 'my-instance'}

4. Включаем вывод стандартного набора метрик:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Enable default Tarantool metrics such as network, memory, operations, etc.:

..  code-block:: lua

>     metrics.enable_default_metrics()

Reviewer: Начиная с версии 0.17.0, для включения метрик по умолчанию и настройке глобальных лейблов рекомендуется использовать `metrics.cfg{}`.

(Что ещё страшнее, в Tarantool 2.11.1 будет рекомендоваться переехать на `box.cfg{metrics = ...}`.)

Reviewer: Не понял, почему данный комментарий был зарезолвлен

Author: Я это поправил ниже, а здесь пропустил.
Теперь везде исправил

            return http_handler(...)
    end)
        :start()
    -- Объявляем вспомогательные функции
    local function http_metrics_handler(request)
        return request:render({ text = json_exporter.export() })
    end

    -- Запускаем базу данных
    box.cfg{
        listen = 3302
        listen = 3301,
    }

Now you can use the HTTP API endpoint ``/metrics`` to collect your metrics
in the Prometheus format. To learn how to obtain custom metrics, check the
:ref:`API reference <metrics-api_reference>`.
    -- Настраиваем модуль метрик
    metrics.set_global_labels{alias = 'my-tnt-app'}
    metrics.enable_default_metrics()

    -- run web server
    local server = http_server.new('0.0.0.0', 8081)
    server:route({path = '/metrics'}, http_metrics_handler)
    server:start()

..  _monitoring-getting_started-http_metrics:

Collect HTTP metrics
--------------------

To enable the collection of HTTP metrics, you need to create a collector first.
To enable the collection of HTTP metrics, you need wrap handler to function
``metrics.http_middleware.v1``:

..  code-block:: lua

    local metrics = require('metrics')
    local httpd = require('http.server').new(ip, port)

>     -- Create a summary collector for latency

Reviewer: Фактически все используют именно summary для сбора HTTP-метрик (в том числе, Grafana dasboard заточен под summary). При этом по умолчанию используется гистограмма (по историческим причинам). Предложение summary как альтернативы и гистограммы как основного может привести к тому, что потом каждый второй будет приходить с вопросами "я настроил HTTP, но у меня нет половины графиков". Кроме того, Getting started как будто тоже хочется держать максимально лаконичным, а мы его расширяем не самой полезной информацией. Мне кажется, любой, кто дойдёт до желания заменить summary на histogram, сможет найти строку со словами "collector" и "summary" и понять, что с ней сделать.

Author: В таком случае по умолчанию должна использоваться summary.
Раз по факту по умолчанию используется histogram, то подразумевается, что это основной вариант. Его я и расписал. Пожалуй тогда укажу, что hisogram используется по умолчанию по историческим причинам.

Reviewer: Если что, под "историческими причинами" я имел в виду вопрос обратной совместимости. Middleware появился тогда, когда summary с квантилями ещё не было.

Reviewer: Если мы преследуем цель "дать людям правильную и лаконичную инструкцию по работе", то надо просто писать пример с использованием summary с минимальным количеством избыточной информации (разве что подчеркнуть, что мы рекомендуем использовать именно summary для интеграции с другими инструментами).

    local httpd = require('http.server').new(ip, port)

    -- Create a summary collector for latency

>     local default_collector = metrics.http_middleware.configure_default_collector('summary')

Reviewer: ```suggestion
    metrics.http_middleware.configure_default_collector('summary')
```

Reviewer 2: Thanks, fixed

    -- Start HTTP routing
    httpd:start()

> .. hint::
>     By default, the ``http_middleware`` uses the ``histogram`` collector for backward compatibility reasons.
>     For collecting HTTP metrics, it's recommended to use ``summary``.

Reviewer 2: ```suggestion
..  note::
    By default, the ``http_middleware`` uses the ``histogram`` collector for backward compatibility reasons.
    To collect HTTP metrics, use the ``summary`` type instead.
```

You can collect all HTTP metrics with a single collector.
If you're using the default
:ref:`Grafana dashboard <monitoring-grafana_dashboard-page>`,
don't change the default collector name.
Otherwise, your metrics won't appear on the charts.

..  _monitoring-getting_started-custom_metric:

Custom metric
-------------

You can create your own metric in two ways, depending on when you need to take measurements:

> - At an arbitrary moment in time

Reviewer: ```suggestion
- At any arbitrary moment of time
```

- At the time of requesting the data collected by the metrics

Let's explore both options.

At a random moment
~~~~~~~~~~~~~~~~~~

> 1. Declare the collector:

Reviewer: ```suggestion
1. Create the collector:
```

..  code-block:: lua

    local response_counter = metrics.counter('response_counter', 'Response counter')

2. Take a measurement at the appropriate place, for example, in an API request handler:

..  code-block:: lua

    local function check_handler(request)
        local label_pairs = {
            path = request.path,
            method = request.method,
        }
        response_counter:inc(1, label_pairs)
        -- ...
    end

> At the time of requesting the data collected by the metrics

Author: Какое?

    end

    local function check_handler(request)
        local label_pairs = {
            path = request.path,
            method = request.method,
        }
        response_counter:inc(1, label_pairs)
        return request:render({ text = 'ok' })
    end

    -- Start the database
    box.cfg{
        listen = 3301,
    }

    -- Configure the metrics module
    metrics.set_global_labels{alias = 'my-tnt-app'}

    -- Run the web server
    local server = http_server.new('0.0.0.0', 8081)
    server:route({path = '/metrics'}, http_metrics_handler)
    server:route({path = '/check'}, check_handler)
    server:start()

Result:

    ..  code-block:: json

    [
      {
        "label_pairs": {
          "path": "/check",
          "method": "GET",
          "alias": "my-tnt-app"
        },
        "timestamp": 1688385933874080,
        "metric_name": "response_counter",
        "value": 1
      }
    ]

..  _monitoring-getting_started-warning:
Warning
-------

You might want to add your own metric. The module allows this, but there are nuances when working with
specific tools.

When adding your custom metric, it's important to ensure that the number of label value combinations is

> kept to a minimum. This is to prevent a "combinatorial explosion" in the database where metric data is

Reviewer: ```suggestion
kept to a minimum. Otherwise, combinatorial explosion may happen in the timeseries database with metrics values
```

stored. Examples of data labels:

- Labels in Prometheus.
- Tags in InfluxDB.

For example, if your company uses InfluxDB for metric collection, you could potentially disrupt the entire
monitoring setup, both for your application and for all other systems within the company. As a result,
monitoring data is likely to be lost.

Let's look at an example:

..  code-block:: lua

    local some_metric = metrics.counter('some', 'Some metric')

    -- THIS IS POSSIBLE
    local function on_value_update(instance_alias)
       some_metric:inc(1, { alias = instance_alias })
    end

    -- THIS IS NOT ALLOWED
    local function on_value_update(customer_id)
       some_metric:inc(1, { customer_id = customer_id })
    end

In the example, there are two versions of the function ``on_value_update``. The top version labels
the data with the cluster node's alias. Since there's a relatively small number of nodes, using
them as labels is feasible. In the second case, an identifier of a record is used. If there are many
records, it's advisable to avoid such situations.

The same principle applies to URLs. Using the entire URL with parameters is not recommended; using a

> URL template or just the name of the command is a better approach. And so on.

Author: Не понял как перефразировать

In essence, when designing custom metrics and selecting labels or tags, it's crucial to opt for a minimal
set of values that can uniquely identify the data without introducing unnecessary complexity or potential

> conflicts with existing metrics and systems.

Author: На какие?
